instrumentation/github.com/labstack/echo/otelecho/echo.go

```diff
@@ -15,90 +15,63 @@
+	echoContextCtxKey echoCtxKey = iota
 )
 
 // Middleware returns echo middleware which will trace incoming requests.
 func Middleware(service string, opts ...Option) echo.MiddlewareFunc {
-	cfg := config{}
-	for _, opt := range opts {
-		opt.apply(&cfg)
-	}
-	if cfg.TracerProvider == nil {
-		cfg.TracerProvider = otel.GetTracerProvider()
-	}
-	tracer := cfg.TracerProvider.Tracer(
-		tracerName,
-		oteltrace.WithInstrumentationVersion(SemVersion()),
-	)
-	if cfg.Propagators == nil {
-		cfg.Propagators = otel.GetTextMapPropagator()
-	}
-
-	if cfg.Skipper == nil {
-		cfg.Skipper = middleware.DefaultSkipper
-	}
-
+	conf := newConfig(opts...)
 	return func(next echo.HandlerFunc) echo.HandlerFunc {
 		return func(c echo.Context) error {
-			if cfg.Skipper(c) {
-				return next(c)
-			}
+			var handlerFunc http.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				c.SetRequest(r)
+				c.SetResponse(echo.NewResponse(w, c.Echo()))
+				err := next(c)
+				if err != nil {
+					span := trace.SpanFromContext(r.Context())
+					if span != nil {
+						span.SetAttributes(attribute.String("echo.error", err.Error()))
+					}
+					// invokes the registered HTTP error handler
+					c.Error(err)
+				}
+			})
 
-			c.Set(tracerKey, tracer)
-			request := c.Request()
-			savedCtx := request.Context()
-			defer func() {
-				request = request.WithContext(savedCtx)
-				c.SetRequest(request)
-			}()
-			ctx := cfg.Propagators.Extract(savedCtx, propagation.HeaderCarrier(request.Header))
-			opts := []oteltrace.SpanStartOption{
-				oteltrace.WithAttributes(semconv.NetAttributesFromHTTPRequest("tcp", request)...),
-				oteltrace.WithAttributes(semconv.EndUserAttributesFromHTTPRequest(request)...),
-				oteltrace.WithAttributes(semconv.HTTPServerAttributesFromHTTPRequest(service, c.Path(), request)...),
-				oteltrace.WithSpanKind(oteltrace.SpanKindServer),
-			}
-			spanName := c.Path()
-			if spanName == "" {
-				spanName = fmt.Sprintf("HTTP %s route not found", request.Method)
+			if conf.routeTagFromPath {
+				handlerFunc = otelhttp.WithRouteTag(c.Path(), handlerFunc)
 			}
 
-			ctx, span := tracer.Start(ctx, spanName, opts...)
-			defer span.End()
-
-			// pass the span through the request context
-			c.SetRequest(request.WithContext(ctx))
-
-			// serve the request to the next middleware
-			err := next(c)
-			if err != nil {
-				span.SetAttributes(attribute.String("echo.error", err.Error()))
-				// invokes the registered HTTP error handler
-				c.Error(err)
-			}
-
-			attrs := semconv.HTTPAttributesFromHTTPStatusCode(c.Response().Status)
-			spanStatus, spanMessage := semconv.SpanStatusFromHTTPStatusCodeAndSpanKind(c.Response().Status, oteltrace.SpanKindServer)
-			span.SetAttributes(attrs...)
-			span.SetStatus(spanStatus, spanMessage)
-
+			ctx := context.WithValue(c.Request().Context(), echoContextCtxKey, c)
+			otelhttp.NewHandler(handlerFunc, service, conf.otelhttpOptions...).
```

Comment:
This is creating a new `otelhttp.Handler` for every request.  That seems like a lot of overhead.  Has this been benchmarked?

Reply:
I've addressed this in my latest changes. It now passes required echo context and next handler through context.